### PR TITLE
correct update-crypto-policies --set is not idempotent

### PR DESCRIPTION
**Pull request by jackmanb1985**
#### Description:
update-crypto-policies --set is not idempotent and will execute on subsequent runs of Ansible. Check and compare the current runtime value of update-crypto-policies by invoking update-crypto-policies --show and registering its output as a variable which can be used as a conditional on the subsequent update-crypto-policies --set

#### Rationale:
Ansible should be idempotent by providing a conditional state to compare against.

### CODE REVIEW

**Review by jan-cerny — CHANGES_REQUESTED** [review]
This PR causes a lot of test scenarios to fail whereas they pass on the current master. Please fix these failures:

```
[jcerny@thinkpad scap-security-guide{pr/8496}]$ python3 tests/test_suite.py rule --libvirt qemu:///system ssgts_rhel8 --remediate-using ansible configure_crypto_policy
Setting console output to log level INFO
INFO - The base image option has not been specified, choosing libvirt-based test environment.
INFO - Logging into /home/jcerny/work/git/scap-security-guide/logs/rule-custom-2022-04-12-1133/test_suite.log
INFO - xccdf_org.ssgproject.content_rule_configure_crypto_policy
INFO - Script config_and_current_same_time.pass.sh using profile (all) OK
INFO - Script config_newer_than_current.fail.sh using profile (all) OK
ERROR - Rule evaluation resulted in fail, instead of expected pass during final stage 
ERROR - The check after remediation failed for rule 'xccdf_org.ssgproject.content_rule_configure_crypto_policy'.
INFO - Script missing_nss_config.fail.sh using profile xccdf_org.ssgproject.content_profile_ospp OK
ERROR - Rule evaluation resulted in fail, instead of expected pass during final stage 
ERROR - The check after remediation failed for rule 'xccdf_org.ssgproject.content_rule_configure_crypto_policy'.
INFO - Script missing_policy.fail.sh using profile xccdf_org.ssgproject.content_profile_ospp OK
ERROR - Rule evaluation resulted in fail, instead of expected pass during final stage 
ERROR - The check after remediation failed for rule 'xccdf_org.ssgproject.content_rule_configure_crypto_policy'.
INFO - Script missing_policy.fail.sh using profile xccdf_org.ssgproject.content_profile_standard OK
ERROR - Rule evaluation resulted in fail, instead of expected pass during final stage 
ERROR - The check after remediation failed for rule 'xccdf_org.ssgproject.content_rule_configure_crypto_policy'.
INFO - Script missing_policy_file.fail.sh using profile xccdf_org.ssgproject.content_profile_ospp OK
ERROR - Rule evaluation resulted in fail, instead of expected pass during final stage 
ERROR - The check after remediation failed for rule 'xccdf_org.ssgproject.content_rule_configure_crypto_policy'.
INFO - Script missing_policy_file.fail.sh using profile xccdf_org.ssgproject.content_profile_standard OK
ERROR - Rule evaluation resulted in fail, instead of expected pass during final stage 
ERROR - The check after remediation failed for rule 'xccdf_org.ssgproject.content_rule_configure_crypto_policy'.
INFO - Script nss_config_as_file.pass.sh using profile xccdf_org.ssgproject.content_profile_ospp OK
INFO - Script nss_config_as_symlink.pass.sh using profile xccdf_org.ssgproject.content_profile_ospp OK
INFO - Script policy_default_cis_l1.pass.sh using profile xccdf_org.ssgproject.content_profile_cis_server_l1 OK
INFO - Script policy_default_cis_l1.pass.sh using profile xccdf_org.ssgproject.content_profile_cis_workstation_l1 OK
INFO - Script policy_default_nosha1_set.pass.sh using profile xccdf_org.ssgproject.content_profile_e8 OK
INFO - Script policy_default_set.pass.sh using profile xccdf_org.ssgproject.content_profile_standard OK
INFO - Script policy_fips_ospp_set.pass.sh using profile xccdf_org.ssgproject.content_profile_ospp OK
INFO - Script policy_future_cis_l2.pass.sh using profile xccdf_org.ssgproject.content_profile_cis OK
INFO - Script policy_future_cis_l2.pass.sh using profile xccdf_org.ssgproject.content_profile_cis_workstation_l2 OK
INFO - Script wrong_policy.fail.sh using profile xccdf_org.ssgproject.content_profile_ospp OK
ERROR - Rule evaluation resulted in fail, instead of expected pass during final stage 
ERROR - The check after remediation failed for rule 'xccdf_org.ssgproject.content_rule_configure_crypto_policy'.
INFO - Script wrong_policy.fail.sh using profile xccdf_org.ssgproject.content_profile_standard OK
ERROR - Rule evaluation resulted in fail, instead of expected pass during final stage 
ERROR - The check after remediation failed for rule 'xccdf_org.ssgproject.content_rule_configure_crypto_policy'.
```

File: linux_os/guide/system/software/integrity/crypto/configure_crypto_policy/ansible/shared.yml [hunk 1]
```diff
@@ -12,5 +12,10 @@
     line: "{{ var_system_crypto_policy }}"
     create: yes
 
+- name: Register Crypto Policy value (runtime)
+  command: /usr/bin/update-crypto-policies --show
+  register: update_crypo_policies_show
```
[the remark this answers is not in the document]
Comment (jackmanb1985) [reply linux_os/guide/system/software/integrity/crypto/configure_crypto_policy/ansible/shared.yml]: updated typo

**Comment by jan-cerny** [comment]
Hi, please remove all "Merge ..." commits from this PR, the "Merge ..." commits can't be accepted.

**Comment by jan-cerny** [comment]
Thanks for the rebase, but the test scenarios are still failing a lot, the test run still has the same output as in the comment above.

For example, the test scenario `config_newer_than_current.fail.sh` modifies mtime on  /etc/crypto-policies/config, but the new Ansible remediation probably doesn't regenerate the configuration, therefore the final stage scan fails on the crypto_policies_updated OVAL test which tests whether the mtime of /etc/crypto-policies/config is less than or equal the mtime of /etc/crypto-policies/state/current. You will probably need to ensure the update-crypto-policies command is exeucted if this file has been modified.

Another example is the test scenario missing_nss_config.fail.sh which removes the file /etc/crypto-policies/back-ends/nss.config and then expects that rule will fail and remediation will regenerate this file. But on the final scan results you can see that the test for presence of this file failed, that means that the remediation didn't create it.

Similarly, the test missing_policy_file.fail.sh removes /etc/crypto-policies/state/current but the remediation didn't create it during the scan.

The pattern I observe is that you can't rely solely on the update-crypto-policies --show to determine whether to set a crypto policy or do nothing. You should follow all the branches of the associated OVAL check https://github.com/ComplianceAsCode/content/blob/08436ce586a70cfaaa22411167d9cdf80fdcb182/linux_os/guide/system/software/integrity/crypto/configure_crypto_policy/oval/shared.xml to figure out how to check for a correct state that requires no action and I guess execute the action in every other situation.

@yuumasato Do you have any additional information?

Also, I have noticed that the corresponding Bash remediation also just runs update-crypto-policies without any preliminary checks. And we usually keep Ansible in line with Bash.

**Comment by yuumasato** [comment]
I think `update-crypto-policyes --is-applied` could be used?/

After modifying the config manually, I get the following:
```
[root@localhost ~]# vim /etc/crypto-policies/config
[root@localhost ~]# update-crypto-policies --is-applied
The configured policy is NOT applied
```

**Comment by jan-cerny** [comment]
@yuumasato That's ingenious!
@jackmanb1985 Would Yuuma's proposal work for you?

**Comment by jackmanb1985** [comment]
@jan-cerny I took the approach from an ansible standpoint when crafting up this logic to firstly understand the desired state and if not currently set in `--show`, then set the value as such in the config file.

As @yuumasato has shown, `--is-applied` appears to come to the same conclusion and based upon your notes on the test suite above and potentially may allow it to pass.  The caveat with both of our examples are if it is not modifying the contents file (which a check is not supposed to), then the test case is always going to fail under these scenarios (because the contents never needs to be updated if ansible identifies it does not need to be changed.)

The test cases sound like they are not fully taking into account ansible best practice under these scenarios?

**Comment by jan-cerny** [comment]
Closing because of inactivity.

**Pull request closed without merging**